added experimental design blog and other minor improvements (#95)
amzn/emukit: _posts/2016-06-09-experimental-design.markdown

---
layout: post
title:  "Experimental design"
img: experimental-design.png
---

> Experimental design addresses the problem of how to collect data points (experiments) to better control certain

Reviewer: "Certain sources of variance" is quite ambiguous. 

I guess the aim of experimental design is to reduce the epistemic uncertainty we have about our system. But let's not use that word because not many people (I don't think) will know what it means...

Author: exactly, I didn't want to use epistemic uncertainty because of that.

sources of variance of a model.

In experimental design the goal to decide in which locations of the input space we should evaluate a function of interest.
In some contexts it is also know as active learning, for instance in image classification problems in which 
more labels need to be collected. 

The are two main ways of doing experimental design. 
 * *Model-free designs:* These designs define rules to spread the experiments as much as possible
across the input domain. Drawing points at random or in a grid are the most naive way of doing so. Other more elaborated approaches are the so-called
[low discrepancy sequences](https://en.wikipedia.org/wiki/Low-discrepancy_sequence) that try to induce some negative correlation in the selected points to spread them evenly. Some examples are 

> Latin designs or Sobol sequences.

Reviewer: ```suggestion
Latin hyper-cube sampling and Sobol sequences.
```

* *Model-based designs:* In these designs a feedback loop is used between some 'optimal' statistical criterion to collect points and a model. In general, the criterion aims to 
reduce some type of variance in the model. The type of optimality refers to which type of uncertainty that is reduced. For instance, a *D-optimal* design aims

> to maximizing the differential Shannon information content of the model parameter estimates; an *I-optimal* design seeks to minimize the average prediction 

Reviewer: ```suggestion
to maximize the differential Shannon information content of the model parameter estimates; an *I-optimal* design seeks to minimize the average prediction 
```

variance over the entire design space. See [[1]](#references-on-experimental-design) for a general review on experimental design of these type with Bayesian modes. 
  

Gaussian processes have a long tradition of being the 'model of choice' for designing experiments [[2](#references-on-experimental-design)]. Next, we explain how
you can use them in Emukit for this purpose. Of course you can generalize these ideas to other models too. 

We start by loading the [Branin function](https://www.sfu.ca/~ssurjano/branin.html). 

> We define the input space to be $$[-5,10]\times [0,15]$$.

Reviewer: Maths doesn't render, try:
```suggestion
We define the input space to be $[-5,10]\times [0,15]$.
```

```python
from emukit.test_functions import branin_function
from emukit.core import ParameterSpace, ContinuousParameter

f, _ = branin_function()
parameter_space = ParameterSpace([ContinuousParameter('x1', -5, 10),
                                  ContinuousParameter('x2', 0, 15)])
```

> Emukit support both, model-free and model-based experimental design strategies. To start, as we don't have any other information about the function we first collect 20 points 

Reviewer: ```suggestion
Emukit supports both model-free and model-based experimental design strategies. To start, as we don't have any other information about the function, we first collect 20 points 
```

using a Latin design.

```python
from emukit.experimental_design.model_free.latin_design import LatinDesign

design = LatinDesign(parameter_space) 
X = design.get_samples(num_data_points = 20)
```

Now we evaluate the function at the selected points and we fit a model with [GPy](https://github.com/SheffieldML/GPy).

```python
from GPy.models import GPRegression
from emukit.model_wrappers import GPyModelWrapper

Y = f(X)
model_gpy = GPRegression(X,Y)
model_emukit = GPyModelWrapper(model_gpy)
```

We can use the model to decide which are the best points to collect using some data collection criteria (that we call acquisition). 

> Here we use the model variance as acquisitions function. It is known that when using Gaussian processes are used, 

Reviewer: ```suggestion
Here we use the model variance as the acquisition function. It is known that when using Gaussian processes, 
```

> selecting points of maximum variance is equivalent to maximize the mutual information between the model and

Reviewer: ```suggestion
selecting points of maximum variance is equivalent to maximizing the mutual information between the model and
```
